Implement rendering pipeline and post-processing

x2048 opened the pull request:
This PR implements rendering pipeline as described in #12411 [Alternative B](https://github.com/minetest/minetest/issues/12411#issuecomment-1149209408) and a minimal post-processing shader on top of the pipeline.

Fixes #12411 and fixes #5446

## To do

This PR is Ready for review

- [x] Define base classes
- [x] Implement simple rendering steps (3D and HUD)
- [x] Implement RenderingCorePlain as a pipeline
- [x] - Rendering to a TextureBuffer
- [x] - Scaling as a pipeline step
- [x] Implement Stereo modes
- [x] - Anaglyph
- [x] - Interlaced
- [x] - ~Page flip~ removed because not supported by current Irrlicht version
- [x] - Side by Side / Top Bottom / Cross eyed
- [x] Handle window resizing in a generic way
- [x] Implement shadow mapping as a pipeline step
- [x] Implement post-processing as a pipeline step
- [x] Implement common pipeline factory for 3D stage
  ~Merge undersampling with all 3d modes~ - postponed to a future PR
- [x] Merge post-processing with all 3d modes
- [x] Move tonemapping to postprocessing
- [x] Optimize use of textures
- [x] Fix spaces vs tabs
- [x] Make it work on GLES2

## How to test

* Choose your favorite stereo mode with or without shaders
* Start the game
* Everything must work correctly

appgurueu committed:
A second stage is born

x2048 committed:
Fix depth buffer in the main rendering pass

x2048 committed:
Normalize draw type before passing to second stage

x2048 committed:
Make sense of depth buffer

x2048 committed:
Minor fixes in second stage code

* Invoke rendering of post-effect
* Rename draw -> applyEffects

x2048 committed:
Initial rendering pipeline framework

x2048 committed:
Add first core rendering steps

x2048 committed:
Use 3D and HUD rendering steps in the core

x2048 committed:
Add trivial pipeline to RenderingCorePlain

x2048 committed:
Remove unnecessary system include

x2048 committed:
Implement undersampling using rendering pipeline

x2048 committed:
Annotate rendering pipeline classes

x2048 committed:
Update sky color in simple core

x2048 committed:
Use clear color in screen target

x2048 committed:
Set screen as render target for 3D and HUD by default

x2048 committed:
Move creation of pipeline to a virtual method

x2048 committed:
Implement anaglyph using render pipeline

x2048 committed:
Remove assignment of 3D and HUB steps to screen

x2048 committed:
Make TextureBuffer work in second stage

x2048 committed:
Switch second stage core to pipeline

x2048 committed:
Add native pipeline step for post effects

x2048 committed:
Replace ColorBuffer with TextureBufferOutput

x2048 committed:
Add <string> header to fix compilation errors on MSVC

x2048 committed:
Use pointer to clear color in RenderTarget

x2048 committed:
Implement pipeline for interlaced 3d mode

x2048 committed:
Add post fx to the pipeline for interlaced 3d mode

x2048 committed:
Implement pipeline for side-by-side 3d modes

x2048 committed:
Move DrawImageStep class to a header

x2048 committed:
Remove dead code

x2048 committed:
Remove RenderingCore::drawPostFx() method

x2048 committed:
Add support for running pipeline with a context

x2048 committed:
Propagate pipeline context to the rendering targets and steps

x2048 committed:
Simplify Draw3D step using pipeline context

x2048 committed:
Simplify DrawHUD step using pipeline context

x2048 committed:
Simplify MapPostFxStep using pipeline context

x2048 committed:
Simplify upscaling step using pipeline context

x2048 committed:
Simplify OffsetCameraStep using pipeline context

x2048 committed:
Simplify SetColorMaskStep using pipeline context

x2048 committed:
Fixup 3b8f975eb set clear color for screen target

x2048 committed:
Rewrite DrawImageStep using pipeline context

x2048 committed:
Fixup remove warning for OffsetCameraStep

x2048 committed:
Simplify PostProcessingStep using pipeline context

x2048 committed:
Use pipeline context in TextureBufferOutput

x2048 committed:
Use pipeline context in ScreenTarget

x2048 committed:
Use pipeline context in ClearDepthBufferTarget

x2048 committed:
Enable pipeline context in TextureBuffer

x2048 committed:
Remove obsolete methods in RenderStep

x2048 committed:
Remove obsolete methods in RenderTarget

x2048 committed:
Pass clear color via pipeline context

x2048 committed:
Initialize interlaced mask in a pipeline step

x2048 committed:
Add support for dynamically sized texture buffers

x2048 committed:
Make all render steps and texture buffers owned by the pipeline

x2048 committed:
Remove dead code in plain rendering core

x2048 committed:
Use scene output to return virtual size for the next frame

x2048 committed:
Remove state from side by side stereo core

* Change DrawImageStep to use relative offset and calculate absolute pos inline
* Remove right frame position and calculations from the core class

x2048 committed:
Remove obsolete methods from RenderingCore

x2048 committed:
Remove obsolete fields from RenderingCore

x2048 committed:
Fixup OffsetCameraStep eye_offset calculation

x2048 committed:
Remove obsolete object references in RenderingCore

x2048 committed:
Remove PipelineState struct.

Use PipelineContext to pass the flags.

x2048 committed:
Inject shadow renderer into steps via pipeline context

x2048 committed:
Move shadow map update to a pipeline step

x2048 committed:
Remove references to basic pipeline steps from RenderingCore

x2048 committed:
Use dynamically allocated pipeline in RenderingCore

x2048 committed:
Break dependency of core.h on pipeline.h

Move the step definitions to plain.h

x2048 committed:
Enable construction of RenderingCore with a custom pipeline

x2048 committed:
Add factory method to create shadow renderer.

This is needed to move creation logic away from RenderingCore

x2048 committed:
Retab newly added source files

x2048 committed:
Remove dummy pipeline creation method

x2048 committed:
Remove TrampolineStep, no longer needed

x2048 committed:
Add DynamicTarget class

Render target that can be redirected later

x2048 committed:
Remove obsolete reset() from RenderSource

x2048 committed:
Add DynamicSource class.

Render source that can be remapped later

x2048 committed:
Add configurable input and output to RenderPipeline to enable nesting

x2048 committed:
Extract postprocessing part of the pipeline into a generic method

x2048 committed:
Extract upscaling setup into a dedicated factory method

Iniquitatis commented on client/shaders/3d_secondstage/opengl_fragment.glsl:
```diff
@@ -0,0 +1,34 @@
+
+#define rendered baseTexture
+#define normalmap normalTexture
+#define depthmap ShadowMapSampler
```
Why `PascalCase`?

x2048 replied:
Can you elaborate the concern?

sfan5 replied:
consistency, I'd assume

x2048 replied:
Both of the inconsistent names are coming from existing code in Minetest, and fixing these names is out of scope for this specific PR. What I can do is introduce new uniforms for post-processing texture names.

Is there a reason why GLSL uses camel case for variables while C++ uses snake case?

kilbith replied:
Sorry nitpickers but I think x2048 has more important things to do at this point than arguing with the variable naming style.

If you're able to make more relevant, in-depth, technical reviews, please do so, otherwise I highly recommend to STFU.

Iniquitatis replied:
@kilbith There's no arguing here, just a question.
Am I able to? Sure, I am. Not sure I'd want to after such an attitude, though.
Besides, _everything_ is relevant to the code, if it's written in the code itself.

Iniquitatis commented on client/shaders/nodes_shader/opengl_vertex.glsl:
```diff
@@ -42,6 +42,13 @@ centroid varying vec2 varTexCoord;
+varying vec3 lightVec;
+varying vec3 tsEyeVec;
+varying vec3 tsLightVec;
+varying float area_enable_parallax;
```
Why `snake_case`?

x2048 committed:
Add scale factor to render pipeline

x2048 committed:
Centralize creation of 3D stage pipeline

x2048 committed:
Simplify calculation of virtual render size

Replace use of dedicated render target with fixed scale

x2048 committed:
Add factory support to generate pipeline outside of rendering core

x2048 committed:
Remove central reference to screen target from rendering core

x2048 committed:
Move creation of plain rendering pipeline outside of rendering core

x2048 committed:
Move creation of anaglyph pipeline outside of rendering core

x2048 committed:
Move creation of interlaced 3d mode pipeline outside of rendering core

x2048 committed:
Move creation of side-by-side 3d pipeline ouside of rendering core

x2048 committed:
Replace direct use of Draw3D in anaglyph pipeline with a factory

x2048 committed:
Rename create3DPipeline -> create3DStage

x2048 committed:
Add 3D stage factory with support for postprocessing

x2048 committed:
Move creation of two-stage pipeline outside of rendering core

x2048 committed:
Deduplicate code in createPipeline

x2048 committed:
Remove obsolete rendering core subclasses

x2048 committed:
Integrate postporcessing with plain pipeline

x2048 committed:
Integrate postprocessing with anaglyph pipeline

x2048 committed:
Integrate postprocessing with interlaced stereo pipeline

x2048 committed:
Integrate postprocesing with side-by-side stereo pipelines

x2048 committed:
Remove secondstage as a 3d mode

x2048 committed:
Remove workaround for legacy shader rendering.

The workaround is not necessary since all rendering with shaders will be postprocessed

x2048 committed:
Correct folder name for second-stage shader

x2048 committed:
Disable debug output in second stage by default

x2048 committed:
Move tonemapping code to post-processing shader

x2048 committed:
Downscale G-buffer when in undersampling mode

x2048 committed:
Reenable z-buffer for wield/item meshes

lhofhansl commented:
Tried. Everything seems to work as before.
@x2048 That's a lot of code to review. Anything you'd like to focus on?

rubenwardy commented on minetest.conf.example:
```diff
@@ -858,8 +858,7 @@
 #    Note that the interlaced mode requires shaders to be enabled.
-#    type: enum values: none, anaglyph, interlaced, topbottom, sidebyside, crossview, pageflip
-# 3d_mode = none
+3d_mode (3D mode) enum none none,anaglyph,interlaced,topbottom,sidebyside,crossview,pageflip
```
Pasted int the wrong file

x2048 replied:
Removed now

x2048 committed:
Add support for adding generic objects to the pipeline

x2048 commented:
@lhofhansl I've made slight changes to the pipeline framework, removing duplicated code here ad there.

lhofhansl commented:
With the latest update I find different behavior:
1. the sun is no longer rendered
2. shadows are suddenly view angle dependent
3. entities are do not have shadows on them anymore

(I'm pretty sure this did not happen when I tested yesterday)

x2048 commented:
Hmm... I can't confirm. Sun and entity self-shadowing can be seen here:
![image](https://user-images.githubusercontent.com/4933697/179370473-eaa802e6-25e7-4fb1-8c38-9b1675914a1d.png)

Area where shadows are rendered is dependent on view angle and it may be noticed if you rotate head fast, when shadow_update_frames is 8 or 16.

lhofhansl commented:
Hmm... Lemme test that again...

lhofhansl commented:
Re-compiled everything. The sun is visible. My mistake.

But the other issues remain. What thew me off is that shadows seem to dependent on the aliasing setting. When aliasing is none, then entities self-shadow, but blocks no longer have shadows. At other aliasing settings blocks have shadows again, but entities don't. So some problems with the pipeline setup I assume.

lhofhansl commented:
Shadow issue remain with latest push (antialiasing affecting whether shadows work or not)

Andrey2470T commented:
@x2048 I don't understand how to test sun rays effect which judging by your words in 'Postprocessing' Minetest Forums topic you added in this PR. I just don't see those rays neither in MTG nor in Devtest games.

x2048 commented:
@Andrey2470T , the godrays are on a [separate branch](https://github.com/x2048/minetest/tree/volumetric_light), to keep this PR scoped.

x2048 commented:
@lhofhansl Can it be a driver problem? I can see all shadows with any AA setting (might be it does not work for me at all?)

x2048 committed:
Fix shader error when tonemapping disabled

lhofhansl commented:
@x2048 possible, but not likely. This is with the current Nvidia driver and only with this change.
Lemme try with Intel as well.

lhofhansl commented:
Nope tested with NVidia and Intel. Definitely something weird. Shadows are off when starting Minetest, then when switching antialiasing (without existing Minetest) shadows show up again.
Again. With both NVidia and Intel.

It is as if shadows are missing from the pipeline and then added when antialiasing is changed.

![screenshot_20220719_161014](https://user-images.githubusercontent.com/446130/179774034-a401202c-248b-4e00-8a3d-0f53835ccbfc.png)
![screenshot_20220719_161053](https://user-images.githubusercontent.com/446130/179774046-3d1db601-87a0-4a4c-8c75-69ef1effe712.png)

Andrey2470T commented:
Yes, they render now. I noted the rays will disappear at once as soon as the sun gets outside the player camera view. I don't know exactly whether this behaviour is correct or needs a fix.

x2048 committed:
Force full alpha on post-processed image

x2048 commented:
@lhofhansl I have tried to reproduce the issue on both Intel HD 4000 and AMD Renoir, with different settings in minetest.conf and vanilla one, nothing.

Shadows are always added [here](https://github.com/x2048/minetest/blob/11273d5f318344e2f86ed4f5ca2b21c1a3a52b1f/src/client/render/factory.cpp#L50) and [here](https://github.com/x2048/minetest/blob/11273d5f318344e2f86ed4f5ca2b21c1a3a52b1f/src/client/shadows/dynamicshadowsrender.cpp#L713). There are only two settings that control it from the pipeline perspective: `enable_shaders` and `enable_dynamic_shadows`.

My guess is that shadow renderer is created, but it cannot allocate space for the textures, because we require more VRAM now. Running minetest with `--verbose` may show the error messages from the shadow renderer.

If that does not help, could you share your minetest.conf? There may be other settings at play.

lhofhansl commented:
I had the shadowmap size set to 4096. When I set it back to 2048 I get the shadows again in all cases.

Before this change, though, I could even set the shadowmap size to 8192 and it would still work.
Why would we use so much more VRAM?

Perhaps you can try with larger shadowmap sizes?

lhofhansl commented:
I tried the "godrays" and it's pretty awesome! Helps with immersion.
Let's track down the (probably) memory issues, so we can merge this.

Then we can do the rays in a followup PR. I'd be excited about that.

rubenwardy commented:
Definitely don't put too much into this PR! And I agree that the godrays are awesome

Whilst I can't be much help on the graphics side (most I've done are postprocessing blur for 2d SFML), I'm planning on reviewing the C++ side of this

lhofhansl commented:
After flying around for a while, even with a shadowmap of a size 2048 the shadows eventually vanish.
@x2048 why would we use more VRAM with this?

x2048 commented:
@lhofhansl We use extra textures for normals and depth in the g buffer. I kept them to enable experimenting with shaders, but we can turn it off if VRAM becomes a concern.

All the textures are downscaled with undersampling.

There is a known issue with light position being suddenly off the shaders. Might be related. I'll try to play around to see whether I can reproduce the shadows issue.

lhofhansl commented on src/client/render/core.cpp:
```diff
@@ -19,104 +19,55 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-			g_settings->getBool("enable_dynamic_shadows")) {
-		shadow_renderer = new ShadowRenderer(device, client);
-	}
+RenderingCore::RenderingCore(IrrlichtDevice *_device, Client *_client, Hud *_hud, 
```
Why do we need two different constructors? When are we using which one?

x2048 replied:
One of them was a legacy, removed now.

lhofhansl commented on src/client/render/plain.cpp:
```diff
@@ -19,58 +19,141 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+{
+	RenderStep *step = new Draw3D();
+	if (g_settings->getBool("enable_shaders")) {
+		RenderPipeline *pipeline = new RenderPipeline();
```
I am somewhat confused about how many pipelines we have.
Initially I thought there is one. But then I see multiple places where we do `new RenderPipeLine()`.

x2048 replied:
I use nested pipeline here to abstract the internals of 3D step+post processing from the stereo modes. I could squeeze it into a flat pipeline, but it would complicate the construction code.

lhofhansl replied:
It's fine this way. I just was confused :)

x2048 committed:
Remove obsolete constructor for RenderingCore

lhofhansl commented:
I traced VRAM usage... In the worst case this is using 950MB of my available 4GB.

x2048 commented:
Enabling debug output will tell you whether SM textures are properly generated (then it's a shader issue) or empty (pipeline / renderer issue).

lhofhansl commented:
OK. Did that (and I already had weather disabled - or rather shadows set to 0.7).
I see the debug textures displayed, so it's a shader issue then.

lhofhansl commented:
Even with the latest push I see shadows missing in about 50% of restarts. Sometime they only show on entities, or own solid blocks, or just on water. As said above, this happens with a NVidia and Intel. So weird, and not obvious from the code.

lhofhansl commented:
@x2048 How can I help here?

x2048 commented:
Not a lot of movement here because I'm looking into a possible issue i 5.6.0

Pantyhose-X commented:
Rendering work is still to be completed
* Reflective water
* Depth of field
* Volunmetric clouds
* Volumetric lighting

lhofhansl commented:
@Pantyhose-X All in different PRs.

lhofhansl commented on client/shaders/nodes_shader/opengl_vertex.glsl:
```diff
@@ -42,6 +42,10 @@ centroid varying vec2 varTexCoord;
+varying vec3 lightVec;
+varying vec3 tsEyeVec;
+varying vec3 tsLightVec;
+varying float area_enable_parallax;
```
This is not used anywhere.

x2048 replied:
Removed

sfan5 commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -488,5 +453,7 @@ void main(void)
-	gl_FragColor = col;
+	gl_FragData[0] = col;
+	gl_FragData[1] = vec4((vNormal + 1.) / 2., DRAW_TYPE / 256.);
+	gl_FragData[2] = vec4((fogDistance-length(eyeVec))/1000, 0,0,0);
```
the additional data is for the sake of the shadows shader, right?

x2048 replied:
This is actually only used for debugging. Removed now

sfan5 commented on src/client/camera.cpp:
```diff
@@ -628,7 +628,7 @@ void Camera::wield(const ItemStack &item)
 {
 	// Clear Z buffer so that the wielded tool stays in front of world geometry
-	m_wieldmgr->getVideoDriver()->clearBuffers(video::ECBF_DEPTH);
+	// m_wieldmgr->getVideoDriver()->clearBuffers(video::ECBF_DEPTH);
```
if this is not needed anymore please remove it whole

x2048 replied:
Removed.

sfan5 commented on src/client/render/anaglyph.h:
```diff
@@ -20,15 +20,51 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	void reset(PipelineContext *context) override {}
+	void activate(PipelineContext *context) override;
+private:
+	video::IVideoDriver *driver;
```
redundant with `context->device->getVideoDriver()`, no?

x2048 replied:
Fully obsolete. Removed.

sfan5 commented on src/client/render/core.cpp:
```diff
@@ -19,104 +19,48 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-	beforeDraw();
-	drawAll();
+	pipeline->reset(&context);
+	pipeline->run(&context);
```
I think it would be better for the methods to take a `PipelineContext&` then, sine this is a temporary object

x2048 replied:
Changed.

sfan5 commented on src/client/render/core.h:
```diff
@@ -26,43 +26,27 @@ class Camera;
 
 public:
-	RenderingCore(IrrlichtDevice *_device, Client *_client, Hud *_hud);
+	RenderingCore(IrrlichtDevice *_device, Client *_client, Hud *_hud, 
+			ShadowRenderer *_shadow_renderer, RenderPipeline *_pipeline, v2f virtual_size_scale);
```
```suggestion
	RenderingCore(IrrlichtDevice *device, Client *client, Hud *hud, 
			ShadowRenderer *shadow_renderer, RenderPipeline *pipeline,
			v2f virtual_size_scale);
```

sfan5 commented on src/client/render/factory.cpp:
```diff
@@ -23,30 +23,63 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	RenderPipeline *pipeline;
+};
+
+void createPipeline(const std::string &stereo_mode, IrrlichtDevice *device, Client *client, Hud *hud, CreatePipelineResult *result);
```
same here for `CreatePipelineResult&`
(doesn't matter that much here though)

x2048 replied:
fixed

sfan5 commented on src/client/render/interlaced.h:
```diff
@@ -21,23 +21,15 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	u8 index;
 };
+
+void populateInterlacedPipeline(RenderPipeline *pipeline, Client *client);
```
btw github says there is no newline at the end here

x2048 replied:
fixed

sfan5 commented on src/client/render/pipeline.cpp:
```diff
@@ -0,0 +1,271 @@
+		m_definitions.resize(index + 1);
+
+	if (m_depth_texture_index == index)
+		m_depth_texture_index = 255;
```
seems to be a placeholder, should be a `#define`

x2048 replied:
Made it a scoped const and added assertions the normal indexes do not overlap with the value.

sfan5 commented on src/client/render/pipeline.h:
```diff
@@ -0,0 +1,391 @@
+
+	video::IVideoDriver *m_driver { nullptr };
+	std::vector<TextureDefinition> m_definitions;
+	core::array<video::ITexture *> m_textures;
```
is there a reason this is not a vector?

x2048 replied:
Compatibility with video::IRenderTarget::setTexture(const core::array<> &, video::ITexture *).

sfan5 commented on src/client/render/pipeline.h:
```diff
@@ -0,0 +1,391 @@
+		video::ECOLOR_FORMAT format;
+	};
+
+	bool ensureTexture(video::ITexture **texture, const TextureDefinition& definition, PipelineContext *context);
```
short documentation comment please

x2048 replied:
added

sfan5 commented on src/client/render/pipeline.h:
```diff
@@ -0,0 +1,391 @@
+	virtual void setRenderTarget(RenderTarget *target) override;
+private:
+	std::vector<RenderStep *> m_pipeline;
+	std::vector< std::unique_ptr<RenderPipelineObject> > m_objects;
```
I think you want a set here in case someone ever calls own twice

x2048 replied:
This is a compile-time problem, and set has worse run-time performance. I can add a debug-only check if necessary.

sfan5 commented on src/client/render/plain.cpp:
```diff
@@ -19,58 +19,141 @@ with this program; if not, write to the Free Software Foundation, Inc.,
-		return;
-	driver->setRenderTarget(lowres, true, true, skycolor);
+	// This is necessary to render shadows for animations correctly
+	context->device->getSceneManager()->getRootSceneNode()->OnAnimate(context->device->getTimer()->getTime());
```
is care taken that this isn't called twice per frame in that case?

x2048 replied:
This is old code, and we can't really do a lot here without rewriting ShadowRenderer.

sfan5 commented on src/client/render/plain.h:
```diff
@@ -20,19 +20,72 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+
+RenderStep *create3DStage();
+RenderStep *create3DStage(Client *client, v2f scale);
+v2f getDownscaleFactor();
```
can be static (private to that file)

sfan5 replied:
still applies

x2048 replied:
Fixed

sfan5 commented on src/client/render/secondstage.cpp:
```diff
@@ -0,0 +1,128 @@
+	material.ZBuffer = true;
+	material.ZWriteEnable = video::EZW_ON;
+	material.MaterialType = shader;
+	for (u32 k = 0; k < MYMIN(video::MATERIAL_MAX_TEXTURES, texture_map.size()); ++k) {
```
don't we have a problem if `texture_map.size()` > `video::MATERIAL_MAX_TEXTURES`?
should this be a `sanity_check`?

x2048 replied:
Changed to assert, because this count is managed compile time.

sfan5 commented on src/client/shadows/dynamicshadowsrender.cpp:
```diff
@@ -707,3 +707,15 @@ std::string ShadowRenderer::readShaderFile(const std::string &path)
+	}
+
+	return nullptr;
+}
```
missing these changes here:
https://github.com/minetest/minetest/commit/eb49b6d85cf9ba59da2321d7fe593986b2f1610f#diff-b9873103ddbea0942c843ccc9b7fbf1d0caf49897cd8bee5d275f3923f44738d

x2048 replied:
These? https://github.com/minetest/minetest/pull/12465/files#diff-9e972db6edd0d7ffc3de3432a4a16e4bf7b885e11b3c8ad84056241a1426457fR714

sfan5 replied:
yes

sfan5 commented:
Tested a bunch of stuff:
* OpenGL, 3d_mode = `{none, sidebyside, anaglyph}` -> works
* OpenGL ES, Irrlicht's shaders -> works
* OpenGL ES, our shaders -> entirely broken, see [debug.txt](https://github.com/minetest/minetest/files/9276898/debug.txt)
  what I'm wondering here is why does Irrlicht complain about D32 not being supported? It's only used in `src/client/render/secondstage.cpp` and I had postprocessing and 3d_mode turned off

sidebyside mode:
This is usable cross-eyed with enough concentration. But there is a bug: the crosshair is not offset so it appears in different places on each eye. Opening the GUI instantly breaks immersion too.
No idea if this is regression, just mentioning it here.

x2048 commented:
For GLES, did you disable shadows?

lhofhansl commented:
@sfan5 Do you get the behavior where sometimes shadows are not displayed?
(I'm playing with view distance = 300, max shadow distance = 300, shadow map size = 8192)

sfan5 commented:
> For GLES, did you disable shadows?

Yes, I always test with a clean config (`./bin/minetest --config <(echo video_driver=ogles2)`)

> @sfan5 Do you get the behavior where sometimes shadows are not displayed?

Gave it a quick test at different settings now and didn't see that bug.

x2048 committed:
Use GLES2-compatible shader syntax

Pantyhose-X commented:
> @Pantyhose-X All in different PRs.

#### I searched for other PRs and couldn't find

> https://github.com/minetest/minetest/pull/12465#issuecomment-1206041576

lhofhansl commented:
@sfan5 and @x2048 . Very strange that I am the only one seeing this. In about 50% of restarts shadows do not work or only work on water. No errors, and only happens with this PR applied.

If you have time could you try one more thing: it's seems to be happening more often when the blocksize is set to 32 (in constants.h). Need to create a new world.

Desour commented on src/client/render/pipeline.h:
```diff
@@ -0,0 +1,391 @@
+	 * @param step reference to the instance.
+	 * @return RenderStep* value of the 'step' parameter.
+	 */
+	template<class T>
+	T *own(T *object)
+	{
+		m_objects.push_back(std::unique_ptr<RenderPipelineObject>(object));
+		return object;
+	}
```
This should probably take a `unique_ptr`:
```suggestion
	template<class T>
	T &own(std::unique_ptr<T> object)
	{
		T &ref = *object;
		m_objects.push_back(std::move(object));
		return ref;
	}
```
(Could still return a raw ptr, if you want.)

You could also add a helper member function for shortness:
```cpp
template<typename T, typename... Args>
T &make_owned(Args&&... args) {
	return own(std::make_unique<T>(std::forward<Args>(args)...));
}
```

Then you have:
```cpp
auto screen = pipeline->own(new ScreenTarget()); // old
auto &screen = pipeline->own(std::make_unique<ScreenTarget>()); // new
auto &screen = pipeline->make_owned<ScreenTarget>(); // shorter
```

x2048 replied:
Added createOwned<T>() and addStep<T>(...) as a shortcut for addStep(createOwned<T>()).

Desour commented on src/client/render/plain.cpp:
```diff
@@ -19,58 +19,141 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+			core::rect<s32>(0, 0, lowres->getSize().Width, lowres->getSize().Height));
+}
+
+RenderStep *create3DStage()
+{
+	return new Draw3D();
+}
+
+RenderStep *create3DStage(Client *client, v2f scale)
```
It would also be nice if these returned a `unique_ptr`.

x2048 replied:
Done

lhofhansl commented:
It would be a shame if we let this bit-rot.

x2048 committed:
Remove unused varying variables from the nodes shader

x2048 committed:
Remove debug code from the shaders

x2048 committed:
Remove commented code

x2048 committed:
Pass PipelineContext by reference instead of pointer

x2048 committed:
Clean up RenderingCore constructor declaration

Co-authored-by: sfan5 <sfan5@live.de>

x2048 committed:
Remove obsolete fields from ClearDepthBufferTarget

x2048 committed:
Pass CreatePipelineResult by reference, not pointer

x2048 committed:
Add final newline to interlaced.h

x2048 committed:
Add name for no-depth-texture marker

Add asserts to avoid overlap of texture indices with the marker.

x2048 committed:
Add doc comment to TextBuffer::ensureTexture

x2048 committed:
Remove unused functions

x2048 committed:
Add static assert for maximum number of textures

x2048 commented:
> sidebyside mode: This is usable cross-eyed with enough concentration. But there is a bug: the crosshair is not offset so it appears in different places on each eye. Opening the GUI instantly breaks immersion too. No idea if this is regression, just mentioning it here.

You need to use 'crossview' for cross-eyed stereo. Works fine with cursor and forms for me (but I can only play for, like, 5 minutes).

x2048 committed:
Add factory helpers to RenderPipeline

Pass unmanaged objects as std::unique_ptr<T> and managed objects as T*

x2048 committed:
Always add post-processing step to the parent pipeline

x2048 committed:
Add support for GLES2.0 depth buffers

x2048 commented:
> * what I'm wondering here is why does Irrlicht complain about D32 not being supported? It's only used in `src/client/render/secondstage.cpp` and I had postprocessing and 3d_mode turned off

@sfan5 D32 is not supported on GL ES 2.0, I've changed it to D16 for that case.

x2048 committed:
Remove unnecessary textures and mappings from the g-buffer/pp-shader

x2048 commented:
@lhofhansl I've tried with MAP_BLOCKSIZE 32, shadows work fine. I don't really know what could cause the shadows to disappear on solid nodes but still be present on e.g. water nodes. It's the same shader code.

lhofhansl commented:
I just tried again. Still happening, only with this PR, with both my graphics cards, and with a vanilla minetest.conf.
Other shader effects work, and in debug I can see the shadow textures, but not the shadows in the actual scene.
As stated, it's hard to debug, because it happens only sometimes, and there are no errors or warnings.

Anything I can do to help debug?

Also, if I'm the only one seeing this, it's obviously not a reason not to merge, I'm just the unlucky one in that case. :(

sfan5 commented on src/client/render/secondstage.cpp:
```diff
@@ -96,7 +96,13 @@ RenderStep *addPostProcessing(RenderPipeline *pipeline, RenderStep *previousStep
+#if ENABLE_GLES
+	if (g_settings->get("video_driver") == "ogles2")
+		depth_format = video::ECF_D16;
+#endif
```
is it possible to check for support and degrade unconditionally?
explicitly checking for gles isn't very clean

x2048 replied:
Thanks for the pointer. Added fallback based on texture support in the driver.

lhofhansl commented:
Here's an example of shadows cast on entities and water, but not solid nodes.

![screenshot_20220822_102952](https://user-images.githubusercontent.com/446130/185983273-cfa3a44b-6916-4d12-b3d7-e5e21e5c1149.png)

kilbith commented:
@lhofhansl Did you try on a different platform ?

lhofhansl commented:
> @lhofhansl Did you try on a different platform ?

All I have is the Linux machine. It's the same one, though, and everything (including other 3D apps) works fine, except for Minetest specifically with this PR applied.

lhofhansl commented:
This is getting weirder and weirder, when I relax the thermal conditions on my laptop (i.e. I let the CPU run a bit faster) the shadows show almost all of the time.

So to me this points to some race condition somewhere, which would also consistent with why it works sometimes but not at other times. Either a multithreading issue (or some race between the CPU the graphics card, although I would know what that would be.)

x2048 committed:
Add graceful fallback for depth texture formats

x2048 committed:
Calibrate distance to wield tool to avoid clipping by environment

x2048 committed:
Add comment for wield camera near plane value

x2048 committed:
Remove extra outputs from shaders

x2048 committed:
Make second stage shader compatible with GLES2

lhofhansl commented:
The problem I've been seeing is still present.

x2048 commented:
Can you try to remove caching from shader constants/uniforms? I noticed previously that it caused glitches.

lhofhansl commented:
Is the caching new in this PR?
Where do I disable that? Do you mean caching in ShaderConstantSetter?

x2048 commented:
No, it's not new. Can you try to remove [these two lines](https://github.com/minetest/minetest/blob/75d88dcae233757d9567c971227b33b0168b5254/src/client/shader.h#L96-L97)?

lhofhansl commented:
Cool. These are the two I found as well, just wanted to confirm we meant the same thing.
If it's not new with this PR I doubt it'll make a different. But I'll try :)

lhofhansl commented:
And strangely enough... It does seem to fix the problem I have been seeing!

Is there some kind of new timing issue, perhaps with the second stage, where the shaders are (somehow) not ready the first time, and we never resend the constants?

lhofhansl commented:
So I'm happy to merge this without the caching. @x2048 do you think that would be OK (is the caching helping much)? Or do do you think we should try to track that down first?

BTW. I tried to check the results of the setters and only cache when the return value indicates success, but that does not solve the problem.

x2048 commented:
It was a long shot, glad that it worked.

I'd like to try and fix this. Caching uniforms was not without a reason.

x2048 commented:
Draft documentation on how to play with the pipeline until a better API is created: https://gist.github.com/x2048/7b398f6e10a0f302971d034931dadfa5

lhofhansl commented:
FYI. I changed the *ShaderSetting template such that I can control for each instance whether it's cached or not.
If all settings in game.cpp, clientenvironment.cpp, shadowScreenQuad.h are cached everything still works fine, so it must be one or more of the *ShaderSettings in shader.cpp itself.

lhofhansl commented:
It's one or more of these in shader.cpp:
```
	CachedPixelShaderSetting<f32, 16, false> m_shadow_view_proj;
	CachedPixelShaderSetting<f32, 3, false> m_light_direction;
	CachedPixelShaderSetting<f32, 1, false> m_texture_res;
	CachedPixelShaderSetting<f32, 1, false> m_shadow_strength;
	CachedPixelShaderSetting<f32, 1, false> m_time_of_day;
	CachedPixelShaderSetting<f32, 1, false> m_shadowfar;
	CachedPixelShaderSetting<f32, 4, false> m_camera_pos;
	CachedPixelShaderSetting<s32, 1, false> m_shadow_texture;
```
(Note I added a "cache" parameter to the template).

x2048 commented:
This is good finding. Did you manage to narrow it down further to a specific setting?

lhofhansl commented:
Not, yet. It seems if I cache any of these I see the effect again. I had limited time yesterday, I'll try some more today.

lhofhansl commented:
Ok. It's these two:
```
	CachedPixelShaderSetting<f32, 16, false> m_shadow_view_proj;
	CachedPixelShaderSetting<f32, 4, false> m_camera_pos;
```

If either of these are cached, shadows sometimes do not display.

x2048 commented:
Is there a branch somewhere that I could pull in? I guess, we can go with two uncached uniforms and solve it later.

lhofhansl commented:
> Is there a branch somewhere that I could pull in? I guess, we can go with two uncached uniforms and solve it later.

Not easily right now, sorry.

lhofhansl commented:
But here's the simple change to shader.h.  And then you can mark the two constants with false.

[shader.diff.txt](https://github.com/minetest/minetest/files/9448319/shader.diff.txt)

x2048 committed:
Improve post-processing step configuration

Use shader info ID instead of material renderer ID as step state
Use tile material and draw type values that avoid collisions with main nodes

x2048 commented:
@lhofhansl Can you try now? I changed what makes the state of PostProcessingStep and which material type is passed when generating the shader.

sfan5 reviewed (approved):
Works on GL and GLES

there's a minor review comment remaining, please address it!

lhofhansl commented:
If we merge as is, we should at least avoid caching the two shader constants I have identified as causing the shadow issues.

x2048 committed:
Make getDownscaleFactor static

x2048 committed:
Disable caching for shadow map VP matrix and light vector

Fixes instability of shadows on certain hardware

Co-authored-by: lhofhansl <lhofhansl@yahoo.com>

x2048 commented:
@lhofhansl I've pushed the changes that disable caching of light space matrix and light vector. Can you verify?

lhofhansl commented on src/client/shader.cpp:
```diff
@@ -214,8 +214,8 @@ class MainShaderConstantSetter : public IShaderConstantSetter
-	CachedPixelShaderSetting<f32, 16> m_shadow_view_proj;
-	CachedPixelShaderSetting<f32, 3> m_light_direction;
+	CachedPixelShaderSetting<f32, 16, false> m_shadow_view_proj;
+	CachedPixelShaderSetting<f32, 3, false> m_light_direction;
```
That's the wrong one. The other one is m_camera_pos.

lhofhansl replied:
```
	CachedPixelShaderSetting<f32, 16, false> m_shadow_view_proj;
	CachedPixelShaderSetting<f32, 4, false> m_camera_pos;
```

x2048 replied:
Done

x2048 committed:
Disable caching on the correct uniforms to fix the shadows issue

lhofhansl commented:
Works.

BTW. The new volumetric light with bloom is pretty awesome. I hope we can get that one shortly after this.
